build.assets/tooling/cmd/difftest/main.go

```diff
@@ -0,0 +1,210 @@
+	ref, err := gitMergeBase(*repoPath, *branch)
+	if err != nil {
+		bail(trace.Wrap(err, "fork point might not exist"))
+	}
+
+	// Get git diff with fork commit
+	changes, err := gitChanges(*repoPath, ref)
+	if err != nil {
+		bail(err)
+	}
+
+	// Get a list of changed files
+	changedFiles, err := getChangedTestFilesFromDiff(changes, *exclude, *include)
+	if err != nil {
+		bail(err)
+	}
+
+	switch command {
+	case "diff":
+		diff(*repoPath, ref, changedFiles, time.Since(start))
+	case "test":
+		test(*repoPath, ref, changedFiles)
+	}
+}
+
+// diff prints diff for debug purposes
+func diff(repoPath string, ref string, changedFiles []string, elapsed time.Duration) {
+	fmt.Printf("Tests changed in %v compared to %v:\n\n", repoPath, ref)
+
+	if len(changedFiles) == 0 {
+		fmt.Println("No changes!")
+		return
+	}
+
+	inspect(repoPath, ref, changedFiles, func(filename string, r CompareResult) {
+		if !r.HasNew() && !r.HasChanged() {
+			return
+		}
+
+		fmt.Printf("- %v:\n", filename)
+
+		for _, n := range r.New {
+			fmt.Printf("  +%v (%v)\n", n.Name, n.RefName)
+		}
+
+		for _, n := range r.Changed {
+			fmt.Printf("  ~%v (%v)\n", n.Name, n.RefName)
+		}
+
+		fmt.Println()
+	})
+
+	fmt.Printf("Time elapsed: %s\n", elapsed)
+}
+
+// test builds and prints go test flags
+func test(repoPath string, ref string, changedFiles []string) {
+	var dirs = make(StringSet)
+	methods := make([]string, 0)
+
+	dollar := "$"
+	if *doubleDollar {
+		dollar = "$$"
+	}
+
+	inspect(repoPath, ref, changedFiles, func(filename string, r CompareResult) {
+		if !r.HasNew() && !r.HasChanged() {
+			return
+		}
+
+		for _, n := range r.New {
+			methods = append(methods, "^"+n.RefName+dollar)
+			dirs[relDirName(filename, *relative)] = struct{}{}
+		}
+
+		if *excludeUpdates {
+			return
+		}
+
+		for _, n := range r.Changed {
+			methods = append(methods, "^"+n.RefName+dollar)
+			dirs[relDirName(filename, *relative)] = struct{}{}
+		}
+	})
+
+	if len(methods) == 0 || len(dirs) == 0 {
+		os.Exit(0)
+	}
+
+	fmt.Printf(`-run "%v"`, strings.Join(methods, "|"))
+	if !*onlyRunFlag {
+		fmt.Printf(" %v", strings.Join(dirs.Keys(), " "))
+	}
+}
+
+// inspect iterates over changes in the repo
+func inspect(repoPath string, ref string, changedFiles []string, fn func(string, CompareResult)) {
+	runners, err := findAllSuiteRunners(repoPath, changedFiles)
+	if err != nil {
+		bail(err)
+	}
+
+	for _, filename := range changedFiles {
+		file, err := gitGetFileFromRevision(repoPath, filename, ref)
+		if err != nil {
+			bail(err)
+		}
+
+		var forkPoint []Method
+
+		if file != "" {
+			forkPoint, err = parseMethodMap(filename, file, nil)
+			if err != nil {
+				bail(err)
+			}
+		}
+
+		head, err := parseMethodMap(path.Join(repoPath, filename), nil, runners)
+		if err != nil {
+			bail(err)
+		}
+
+		r := compare(forkPoint, head)
+
+		fn(filename, r)
+	}
+}
+
+func relDirName(filename string, relative string) string {
+	f := filename
+
+	if strings.HasPrefix(f, relative+"/") {
```

Comment:
Use `filepath.Rel` here instead? I'm always weary of manually working with filepaths instead of using the stdlib tools provide.